src/config.rs

```diff
@@ -334,6 +334,90 @@ impl Context {
+    pub(crate) async fn is_self_addr(&self, addr: &str) -> Result<bool> {
+        let addr = addr_normalize(addr).to_lowercase();
+
+        // The addresses we get here are already normalized and lowercase
+        Ok(
+            self.get_primary_self_addr().await.ok().as_deref() == Some(&addr)
+                || self.get_secondary_self_addrs().await?.contains(&addr),
+        )
+    }
+
+    /// Sets `primary_new` as the new primary self address and saves the old
+    /// primary address (if exists) as a secondary address.
+    pub(crate) async fn set_primary_self_addr(&self, primary_new: &str) -> Result<()> {
+        let primary_new = addr_normalize(primary_new).to_lowercase();
+
+        if self.get_primary_self_addr().await.ok().as_deref() != Some(&primary_new) {
+            // Make sure that all server UIDs will be resynced
+            self.sql
+                .execute("DELETE FROM imap_sync;", paramsv![])
+                .await?;
+        }
+
+        // add old primary address (if exists) to secondary addresses
+        let mut secondary_addrs = self.get_all_self_addrs().await?;
+
+        for a in secondary_addrs.iter_mut() {
+            *a = addr_normalize(a).to_lowercase();
+        }
+
+        // never store a primary address also as a secondary
+        secondary_addrs.retain(|a| a != &primary_new);
+        self.set_config(
+            Config::SecondaryAddrs,
+            Some(secondary_addrs.join(" ").as_str()),
+        )
+        .await?;
+
+        self.set_config(Config::ConfiguredAddr, Some(&primary_new))
+            .await?;
+
+        Ok(())
+    }
+
+    /// Returns all primary and secondary self addresses.
+    pub(crate) async fn get_all_self_addrs(&self) -> Result<Vec<String>> {
+        let mut ret = Vec::new();
+
+        ret.extend(self.get_primary_self_addr().await.into_iter());
+        ret.extend(self.get_secondary_self_addrs().await?.into_iter());
+
+        Ok(ret)
+    }
+
+    /// Returns all secondary self addresses.
+    ///
+    /// The addresses are already normalized and lowercased in the database.
+    pub(crate) async fn get_secondary_self_addrs(&self) -> Result<Vec<String>> {
+        let secondary_addrs = self
+            .get_config(Config::SecondaryAddrs)
+            .await?
+            .unwrap_or_default();
+        Ok(secondary_addrs
+            .split_ascii_whitespace()
+            .map(|s| s.to_string())
+            .collect())
+    }
+
+    /// Returns the primary self address.
+    ///
+    /// Normalizes and lowercases the address since the ConfiguredAddr
+    /// may not be lowercased (to start lowercasing it now, we would
+    /// need a db migration - which we can do at some point in the future)
+    pub async fn get_primary_self_addr(&self) -> Result<String> {
+        let ret = self
+            .get_config(Config::ConfiguredAddr)
+            .await?
+            .context("No self addr configured")?;
```

Comment:
I think `get_primary_self_addr` should return `Result<Option<String>>` so when no address is configured it's not an error. Otherwise all the code that calls `get_primary_self_addr` uses `ok()` immediately afterwards and basically ignores database errors in `get_config`, treating them the same way as unconfigured address.

Or, if "No self addr configured" is actually an error and the caller should not call `get_primary_self_addr` when account is not configured, then this error should be bubbled up with `?`, not ignored with `ok()`.

Reply:
Actually I just moved it and renamed it from `get_configured_addr()`, which in this case means that I don't have an opinion on this, but ping @hpk42, I guess you will have an opinion?

Reply:
For now I just pushed [e58f5d3](https://github.com/deltachat/deltachat-core-rust/pull/2896/commits/e58f5d312ad8864324951feb7b1392d192e33a98) so that db errors are not ignored anymore (there are 14 uses of get_configured_addr() in the code where we want to return an Err in case it's unconfigured, so I didn't want to change it to return `Result<Option<_>>`).